src/main/java/pingis/controllers/LiveChallengeController.java

```diff
@@ -261,16 +291,75 @@ public RedirectView playArcade(RedirectAttributes redirectAttributes,
       redirectAttributes.addFlashAttribute("minLength", Integer.MAX_VALUE);
       logger.info("User has test turn. Redirecting to /newtaskpair.");
       return new RedirectView("/newtaskpair");
+
     } else {
       Task implTask = gameplayService.getRandomImplementationTask(challenge);
       Task testTask = taskService.getCorrespondingTask(implTask);
-      redirectAttributes.addAttribute("taskId", implTask.getId());
-      redirectAttributes.addAttribute("testTaskInstanceId",
-          taskInstanceService.getByTaskAndUser(testTask, testTask.getAuthor()).getId());
-      logger.info("User has implementation turn. Redirecting to /newTaskInstance");
-      return new RedirectView("/newTaskInstance");
+      TaskInstance testTaskInstance =
+          taskInstanceService.getByTaskAndUser(testTask, testTask.getAuthor());
+      logger.info("User has implementation turn.");
+      return newTaskInstance(implTask, testTaskInstance, redirectAttributes);
+    }
+  }
+
+  private RedirectView playPractice(RedirectAttributes redirectAttributes, Challenge challenge) {
+    // check for unfinished instances
+    User player = userService.getCurrentUser();
+    logger.info("User: " + player.getName());
+
+    if (challenge.getAuthor().equals(player)
+        || (challenge.getSecondPlayer() != null
+            && challenge.getSecondPlayer().equals(player))) {
+      redirectAttributes.addFlashAttribute("message", "Cannot re-do your own live challenge");
+      return new RedirectView("/error");
+    }
+
+    Optional<TaskInstance> unfinished = player.getTaskInstances().stream()
+        .filter(e -> e.getChallenge().equals(challenge))
+        .filter(e -> e.getStatus() == CodeStatus.IN_PROGRESS)
+        .findFirst();
```

Comment:
Repeating an earlier comment: this might make sense as a separate `TaskInstanceService` method.